build(bazel): use workspace status command for AIO version stamping

kormide opened the pull request:
@devversion @josephperrott @gregmagolan 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

AIO navigation map requires local git sha info to build but this fails in the exec root.

## What is the new behavior?

Dgeni package now uses Bazel stamps.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I think this is the only case where git local commands are used under the execroot.

kormide committed:
build(bazel): use workspace status command for AIO version stamping

Fixes a bug where the navigation map did not have the correct SHAs.

gregmagolan commented on aio/BUILD.bazel:
```diff
@@ -37,6 +37,7 @@ npm_package_bin(
     env = {
         "BAZEL_DGENI_OUTPUT_PATH": "$(@D)",
         "GIT_BIN": "$(GIT_BIN_PATH)",
+        "VOLATILE_STATUS": "$(BINDIR)/../../volatile-status.txt",
```
nit: perhaps `BAZEL_VOLATILE_STATUS_FILE`? That is the naming convention we used in bazel-lib that rules_js inherits https://github.com/aspect-build/bazel-lib/blob/951e1ec9ad5553a00d6b6c135f6c490d9344cc62/lib/private/run_binary.bzl#L65

gregmagolan commented on aio/tools/transforms/angular-base-package/services/bazelStampedVersionInfo.js:
```diff
@@ -0,0 +1,37 @@
+};
+
+function getStampedVariables() {
+  const volatileStatus = fs.readFileSync(process.env.BAZEL_VOLATILE_STATUS_FILE, 'utf8');
```
LGTM.
I have a non-public API rule that does something similar in rules_js if you're curious to see another implementation https://github.com/aspect-build/rules_js/blob/770458840e7b1b88711efc3cb5f1ee63f1206d76/js/private/expand_template.js#L18. That bit originally came from something Alex wrote.
The corresponding rule
https://github.com/aspect-build/rules_js/blob/main/js/private/expand_template.bzl

kormide commented on aio/BUILD.bazel:
```diff
@@ -36,6 +36,7 @@ npm_package_bin(
     data = ["//aio/tools/transforms/angular.io-package"],
     env = {
         "BAZEL_DGENI_OUTPUT_PATH": "$(@D)",
+        "BAZEL_VOLATILE_STATUS_FILE": "$(BINDIR)/../../volatile-status.txt",
```
Oh? I thought only `pkg_npm` supports stamping.

gregmagolan replied:
https://github.com/bazelbuild/rules_nodejs/blob/b02128b178324adbaeeb9195fd24ba207a554605/internal/node/npm_package_bin.bzl#L28

gregmagolan replied:
It looks like the docs don't have it because we didn't add it to the macro docstring

gregmagolan replied:
The environment variables set automatically are `BAZEL_VERSION_FILE` and `BAZEL_INFO_FILE`: https://github.com/bazelbuild/rules_nodejs/blob/b02128b178324adbaeeb9195fd24ba207a554605/internal/providers/node_runtime_deps_info.bzl#L161

gregmagolan replied:
where BAZEL_VERSION_FILE is your volatile status file and BAZEL_INFO_FILE your stable

gregmagolan replied:
docs are a little hard to find for the attribute. Here is the source of truth https://github.com/bazelbuild/rules_nodejs/blob/b02128b178324adbaeeb9195fd24ba207a554605/nodejs/private/providers/stamp_setting_info.bzl#L10

gregmagolan replied:
You probably want `stamp = "@rules_nodejs//nodejs/stamp:use_stamp_flag"`

kormide replied:
It looks like this `stamp` target doesn't get published to the `build_bazel_rules_nodejs` repository, however the environment variable is available even without adding `stamp`.

gregmagolan replied:
It should be in the `@rules_nodejs` core repository which `build_bazel_rules_nodejs` depends on. Is the `@rules_nodejs//nodejs/stamp:use_stamp_flag` target not valid?

kormide replied:
Oh I thought that was a typo and tried `@build_bazel_rules_nodjes` instead. In any case, it doesn't seem to be needed. Should I just leave it out?

kormide replied:
Well, now the variable no longer seems to be available. I'm not sure what happened. Maybe my bazel cache didn't clear and it was using the old variable I had.

Anyway, I can add that stamp attribute, but I'm not getting the `BAZEL_VERSION_FILE` environment variable.

kormide replied:
Using `@rules_nodejs//nodejs/stamp:always` works, which I think is what we want since the build is always stamped.

kormide replied:
Eugh, sorry for the confusion. The reason why it was working earlier, and then not, was because I didn't pass `--config=release` the second time. The `stamp` value you posted works. `@rules_nodejs//nodejs/stamp:always` regardless of whether I pass a config, which I think is what I actually want in this case.

devversion commented on aio/tools/transforms/angular-base-package/services/bazelStampedVersionInfo.js:
```diff
@@ -0,0 +1,40 @@
+  currentVersion.full = `${versionInfo.currentVersion.version}+sha.${properties.BUILD_SCM_ABBREV_HASH}`;
+  currentVersion.commitSHA = properties.BUILD_SCM_HASH;
+  
+  return {...versionInfo, currentVersion};
```
Have you checked if other fields need help too? like the `previousVersions` field?


https://github.com/angular/dgeni-packages/blob/c72834c853cb4d3756d5f4d64f2637af30f2465f/git/services/versionInfo.js#L137

kormide replied:
There may be other places where it *could* be used but isn't needed in AIO. When it's missing it's obvious in the output because SHAs/versions are missing.

Do you want me to take a deeper look and fix any case that might occur?

devversion replied:
I wonder- should we just limit to explicit properties instead of merging with `versionInfo` then? that way we can guarantee that the explicit fields are correct under Bazel?

kormide replied:
I was planning to do that, but `versionInfo` (or `currentVersion`?) is dumped in it's entirety into a template so I wanted to preserve the structure of it.

devversion replied:
Into a template? are we not consuming only certain parts e.g. like `.sha`?

kormide replied:
Yeah this is the spot: https://github.com/angular/angular/blob/main/aio/tools/transforms/angular.io-package/processors/processNavigationMap.js#L30

kormide replied:
"Template" wasn't the right term, but it's hardcoded into the dgeni output and the app uses it to form the navigation map.

devversion replied:
Yes, but we could limit that API contract since we only access explicit properties: https://cs.github.com/angular/angular/blob/07017a7bd30c6bb7bd1b94a9fd1b112ee68c9ced/aio/src/app/navigation/navigation.model.ts#L18-L45.

devversion commented on aio/BUILD.bazel:
```diff
@@ -40,6 +40,7 @@ npm_package_bin(
     },
     exec_properties = ENABLE_NETWORK,
     output_dir = True,
+    stamp = "@rules_nodejs//nodejs/stamp:always",
```
What happens if `--stamp` is not set OOC?

kormide replied:
It still looks at the volatile-status file without `--stamp`. I'm not using `stamp` with any of the AIO build configurations.

devversion replied:
But won't you miss running the workspace status command? https://github.com/aspect-build/angular/blob/aio-bazel-local-git/.bazelrc#L52

kormide replied:
Hm, you'd think so... Let me figure out why exactly this is working.

devversion replied:
I believe you are just getting the old workspace status volatile status- which is still accessible but just not updated.

kormide replied:
You're right, it does just look at the old volatile status and we need a stamped configuration to properly populate it. However, setting `@rules_nodejs//nodejs/stamp:always` causes rules_nodejs to provide the `process.env.BAZEL_VERSION_FILE` env var which is useful to have so the dgeni processor works. If it's stale or not filled out, then in the worst case the locally serve aio app is just missing the SHA of the source tree in the footer.

kormide commented:
Updated and simplified. I no longer override version info in a dgeni service but instead provide a service that can access stamp vars. Rather than dumping the full semver version info into the app, I'm only selectively adding the fields it needs.
